### ISSUE TITLE
[Freetype] TextureFilters using mipmaps with Japanese characters

### ISSUE THREAD

**Issue by clementhely** [libgdx/libgdx#3648]
Trying to integrate Japanese language in our games we ran into problems regarding fonts.

We use `FreetypeFontLoader` to load our fonts in different sizes from `.ttf` files. As some text is heavily scaled up or down, we use mipmap filters to create the `BitmapFont`s.

We use the following filters:
```java
fontParam.fontParameters.minFilter = Texture.TextureFilter.MipMapLinearLinear;
fontParam.fontParameters.magFilter = Texture.TextureFilter.Linear;
fontParam.fontParameters.genMipMaps = true;
```

However, using mipmaps, some Japanese characters are not properly displayed. They are actually drawn, but made almost transparent, as you can see with the text block at the bottom on these screenshots:
![nexus9-with-mipmap](https://cloud.githubusercontent.com/assets/3869204/11870907/a5c0edb6-a4c3-11e5-953d-98268bfa5f73.png)

![xperia-tipo-with-mipmap](https://cloud.githubusercontent.com/assets/3869204/11870910/aaeb3f4e-a4c3-11e5-8941-8a7e285d55e0.png)

Removing the previous block of code and using only the default filters or `TextureFilter.Linear` fixes the problem: every character is correctly drawn, with its right colour, since these filters do not use mipmaps, I assume. Now the problem is that scaled text does not look as sharp as it used to. On larger screens it is quite pixelated, on smaller screens it is pixelated as well and some of it is unreadable (like the score and timer at the top of the screen):
![nexus9-without-mipmap](https://cloud.githubusercontent.com/assets/3869204/11871027/754a9b18-a4c4-11e5-8586-8f1b0464ddaa.png)

![xperia-tipo-without-mipmap](https://cloud.githubusercontent.com/assets/3869204/11871031/7b31b1c4-a4c4-11e5-851c-49204a8a06de.png)

Note that the original issue is something which only happens with Japanese characters, we have no problems with numbers, punctuation or Latin characters (we have not tried with other family of languages, though). Could that be an issue related to the use of `incremental` on the `FreeTypeFontParameter`? Or would it be a problem directly related to the mipmap `TextureFilter`s?

**Also not that we are not using regular `Label`s as they do not support scaling. We extended `Label` to create a widget which calls the corresponding `setFontScale` methods whenever it is scaled.**

**Comment by Darkyenus** [comment]
This isn't a solution, just a workaround, but maybe those two are the same:

I am not sure if it is possible to scale fonts well, even if using mipmaps, that is why distance-field fonts are used. But since you need to generate fonts incrementally, how about generating the exact size glyphs you need (and then using nearest filtering)? You would need more texture space for glyphs, but you wouldn't have to use mipmaps, so in some cases it would actually save space.

**Comment by clementhely** [comment]
The thing is we are developing for Android, so we have to generate bitmap fonts on the fly, otherwise we would need hundreds of bitmaps in order to support all existing screen sizes and densities.

At the beginning we were using bitmap fonts directly, with distance-field, but most text looked a bit blurry when scaled. That's why we started using FreetypeFontLoader with these filters and managed to make text look quite sharp.

**Comment by Darkyenus** [comment]
And generating glyphs in exactly the size that you need on the fly, without scaling, without DF? That should work well IMO.

**Comment by intrigus** [comment]
Which gdx version do you use? There were some improvements for FreeType.

**Comment by clementhely** [comment]
@intrigus We use v1.7.0 for now. I saw yesterday that 1.7.2 brought a lot of improvements related to FreeType, so I gave it a try, but it still doesn't solve it. Text is more visible when made very small, but still not legible. And still quite aliased on bigger screens. And playing around with filters still has the same effect.

@Darkyenus Thank you for the suggestion, I tried that out and it does seem to give the best result. :+1:  It does cause some performance issues apparently, but we could definitely solve that. I think for now we'll stick to our initial workaround as we have a few problems with text size across devices when generating glyphs at their exact size. This will need quite a lot of work in terms of how we handle `Label`s.


As for the initial issue, I think it still remains a problem, even if this is partly caused by the way we decided to manage fonts and labels. In my opinion, `TextureFilter`s should not make part of the text unreadable.

**Comment by Darkyenus** [comment]
Well, the thing is, that there is no way to scale + filter ordinary textures so they don't lose quality and sharpness. Nearest is pixely and linear is blurry and somewhat pixely. Mipmaps don't help here much, because they just "preapply" linear filtering, so it does not keep crisp antialiased edges, but just "blurs" everything. DFF solve this problem by creating the crisp antialiased edge at runtime, when the exact scale is known. But they take too long to generate. What would solve this (but would be a huge overkill and very difficult with current code) would be to use "vector fonts" (don't draw them as bitmaps on rectangles but as ordinary GL trinagles/polygons) and then use some sort of AA/multisampling.

TL;DR: It is complicated, I usually just stick to generating "pixel perfect" fonts, because they are simple and look great

**Comment by xoppa** [comment]
Is this actually an issue of libgdx, or are you not satisfied with the texture sampling you've chosen, or are you perhaps just asking for help? Note you don't have to use tri-linear filtering if that doesnt fit your needs, e.g. you could also use `MipMapNearestNearest` if you don't like the non-pot "averaging", or `MipMapLinearNearest` if you only want to use bi-linear filtering on one of your mipmap levels. You could even supply your own mipmap images to match your exact needs. If you verified that this is actually an issue of libgdx then please include the required information. Otherwise please close this issue. https://github.com/libgdx/libgdx/wiki/More-Info.

**Comment by NathanSweet** [comment]
It is nice to see someone using on the fly glyphs!

I agree with Darkyenus, filtering will always affect the crispness of your images. We generate FreeType font mipmaps the usual way, by scaling. We don't render the text with FreeType multiple times at each level. Maybe that could be done (at significant effort, PR welcome!), but I'm not sure we can guarantee the rendered glyph match the scaled size exactly.

Also I agree with Xoppa, it's not clear if this issue is really a problem in libgdx.

BTW, this is off topic, but Label can be [scaled](https://github.com/libgdx/libgdx/wiki/Scene2d.ui#rotation-and-scale) (as can any group of actors).

**Comment by clementhely** [comment]
Thank you for your help and advice! :+1: 

I guess it's indeed not an issue in libGDX. It was just me misunderstanding the purpose and usage of filtering! I'll close this issue now and will try to rework the way we display text in our app, to avoid having too much scaling happening.
